src/btree/traversal.ts

```diff
@@ -0,0 +1,100 @@
+import { BPTree, ReferencedValue } from "./bptree";
+import { BPTreeNode, MemoryPointer } from "./node";
+
+export type TraversalRecord = {
+	node: BPTreeNode;
+	index: number;
+	pointer: MemoryPointer;
+};
+
+export class TraversalIterator {
+	private tree: BPTree;
+	private key: ReferencedValue;
+	private records: TraversalRecord[];
+
+	constructor(tree: BPTree, key: ReferencedValue) {
+		this.tree = tree;
+		this.key = key;
+		this.records = []; // note this works iff all records are non-empty
+	}
+
+	async init(): Promise<boolean> {
+		const rootResponse = await this.tree.root();
+
+		if (rootResponse.rootNode === null) {
+			return false;
+		}
+
+		const root = rootResponse.rootNode;
+		const offset = rootResponse.pointer;
+		console.log("root: ", root.keys.length);
+		console.log("rootnode: ", offset.offset);
+
+		const path = await this.tree.traverse(this.key, root, offset);
+		this.records = path;
+		console.log("ineedaonedance: ", this.records[0].index);
+		return true;
+	}
+
+	getKey(): ReferencedValue {
+		return this.records[0].node.keys[this.records[0].index];
+	}
+
+	getPointer(): MemoryPointer {
+		return this.records[0].node.pointer(this.records[0].index);
+	}
+
+	async increment(i: number, delta: number): Promise<boolean> {
+		if (i === this.records.length) {
+			return false;
+		}
+
+		this.records[i].index += delta;
+		const rolloverLeft = this.records[i].index < 0;
+		const rolloverRight =
+			this.records[i].index >= this.records[i].node.numPointers();
+
+		if (rolloverLeft || rolloverRight) {
+			if (!this.increment(i + 1, delta)) {
+				return false;
+			}
+
+			try {
+				const node = await this.tree.readNode(
```

Comment:
It would be fine to let this error propagate, it doesn't have to be caught.

Reply:
That was my initial intuition, but in order for the iterator to properly halt, we need to cast the error as a `false` value. 

The error that propagates is an undefined indexing error, which means when it reaches the first `out of bounds` error, we can cast it as `false` and essentially halt the iterator.

Reply:
Which test is this during? `readNode()` should only error if the node itself can't be read due to network issues?

Reply:
> Which test is this during? `readNode()` should only error if the node itself can't be read due to network issues?

The iterator test cases in `bptree.test.ts` errs when we remove the `try {} catch {}` block. 

It errs: 
```sh
TypeError: Cannot read properties of undefined (reading 'node')

      58 |
      59 |                      const node = await this.tree.readNode(
    > 60 |                              this.records[i + 1].node.pointer(this.records[i + 1].index)
         |                                                  ^
      61 |                      );
      62 |
      63 |                      // propagate the rollover
```

This is because `this.records[i + 1]` is undefined. I removed the block and caught it earlier on in `increment`